src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -119,6 +164,17 @@ private void ProduceTags(TaggerContext<TTag> context, DocumentSnapshotSpan spanT
                 var requestedSpan = spanToTag.SnapshotSpan;
                 var editorSnapshot = requestedSpan.Snapshot;
 
+                // Try to get the text snapshot that these diagnostics were created against.
+                // This may fail if this tagger was created *after* the notification for the
+                // diagnostics was already issued.  That's ok.  We'll take the spans as reported
+                // and apply them directly to the snapshot we have.  Either no new changes will
+                // have happened, and these spans will be accurate, or a change will happen
+                // and we'll hear about and it update the spans shortly to the right position.
+                _diagnosticIdToTextSnapshot.TryGetValue(id, out var diagnosticSnapshot);
```

Comment:
Is it possible for a file to be closed/reopened so quickly that _diagnosticIdToTextSnapshot will still have a valid ID but a snapshot from the previous buffer? If so, all the mapping stuff below will crash.

(Going to flag this "changes required" to get a confirmation on this. If it's fine then obviously no changes required.)

Reply:
Yes, that's a possible problem.  Question about this.  If i switch to ITextImage: 

1. how do i validate that it's for the same 'buffer' as the buffer i currently have?
2. how do i use that to map spans forward?

Reply:
Ah, i see you answered that below.  Thanks!

Reply:
So not sure which approach you're doing now. Maybe you have to stay using ITextSnapshot but then you can just compare buffers. Maybe it's best to do that and we can research more efficient ways than me sidetracking you down a crazy long path.

Reply:
> So not sure which approach you're doing now. Maybe you have to stay using ITextSnapshot but then you can just compare buffers. Maybe it's best to do that 

Yup.  THat's the approach i went with and have included in the PR.